### DISCUSSION TITLE
FastAPI DB Context for service layer

### DISCUSSION THREAD

**Discussion by Brellio** [fastapi/fastapi#6616] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
```

### Description

My current set up is FastAPI, SqlAlchhemy & Alembic, although i am writing the raw SQL myself, pydantic etc, pretty straight forward.

I am trying to implement a DDD type pattern where i have a repository layer which handles all DB actions, and a Service Layer which handles all business logic. 

I have basic CRUD routes which communicate directly to my repository layer and all is working. In these methods I am able to successfully use the DB dependency injection. See example code below:

**Dependencies**
```
def get_database(request: Request) -> Database:
    return request.app.state._db
```

```
def get_repository(Repo_type: Type[BaseRepository]) -> Callable:
    def get_repo(db: Database = Depends(get_database)) -> Type[BaseRepository]:
        return Repo_type(db)
    return get_repo
```

**Example GET by ID Route**
```
@router.get("/{id}/", response_model=TablePub, name="Get Table by id")
async def get_table_by_id(
    id: UUID, table_repo: TableRepository = Depends(get_repository(TableRepository))
) -> TableInDB:
    table = await table_repo.get_table_by_id(id=id)
    if not table:
        raise HTTPException(status_code=HTTP_404_NOT_FOUND, detail="No Table found with that id.")
    return table
```

**Corresponding Repository**
```
from databases import Database

class BaseRepository:
    def __init__(self, db: Database) -> None:
        self.db = db

class TableRepository(BaseRepository):

    async def get_table_by_id(self, *, id: UUID) -> TableInDB:
        table = await self.db.fetch_one(
            query=GET_TABLE_BY_ID_QUERY, 
            values={"id": id},
        )
        if not table:
            return None
        return TableInDB(**table)
```

Now I want to start doing some more complex operations and want to add a service layer to house all of the business logic.

What is the correct way to structure this so that i can reuse the repositories that i have already written? For example, i want to return all Sales for a Table, but i need to get the table number from the DB first before i can query the Sales Table. The route requires table_id to be passed in as a param -> service layer, where i fetch the table by ID (Using existing repo) -> from that object, get the table number, then do a request to an external API that requires the table number as a param.

What I have so far:

**Route**
```
@router.get("/{table_id}", response_model=SalesPub, name="Get Sale Entries by table id")
async def get_sales_by_table_id(
    table_id: UUID = Path(..., title="ID of the Table to get Sales Entries for")):
    response = await SalesService.get_sales_from_external_API(table_id=table_id)
    return response
```

**Service Layer 'SalesService'**
```
async def get_sales_from_external_API(
        table_id: UUID, 
        table_repo: TableRepository = Depends(get_repository(TableRepository))
        ) -> TableInDB:
        table_data = await table_repo.get_table_by_id(id=table_id)
        if table_data is None:
            logger.info(f"No table with id:{table_id} could not be found")
        table_number = table_data.number
        client_id = table_data.client_id
        
        sales = await salesGateway.call_external_API(table_number, client_id)
        return sales
```
The code brakes here `table_data = await table_repo.get_table_by_id(id=table_id)`

With an error` AttributeError: 'Depends' object has no attribute 'get_table_by_id'`

What i don't understand is that the code is almost identical to the route method that can get the table by ID? The depends object TableRepository does have a get_table_by_id method. What is it that i'm doing incorrectly, and is this the best way to split up business logic from database actions?

Thanks in advance

### Operating System

macOS

### FastAPI Version

0.55.1

### Python Version

3.8

**Comment by negadive** [comment]
you didnt put `Depends(get_repository(TableRepository))` inside route function params, so its returning instance of `Depends` instead of processing the `get_repository(TableRepository)` and returning instance of `TableRepository `
